### PR TITLE
Add new `Table` block (MVP supporting local data only)

### PR DESCRIPTION
**Pull request by yusufkinatas**
## 🌟 What is the purpose of this PR?

This is the first PR for the new Table block. It will be follow up PRs implementing new features like query building, linking data, filtering, sorting etc. This PR just introduces the most basic functionality and the usage of `glide-data-grid`.

Here is the list of the supported functionalities for now

- local columns
- local rows
- adding colums / rows
- editing cells
- updating table title
- readonly mode

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1203179653350866/f) _(internal)_

## 🐾 Next steps

Implement rest of the features for Table block

## ❓ How to test this?

1.  Run `table-v2` block on your local
2. Confirm that it works as expected

## 📹 Demo

⚠️ Showing copy-paste using keyboard shortcuts at the beginning of the demo

https://user-images.githubusercontent.com/39553853/223970783-8c182290-6d1a-4027-9e56-214b01e6bb13.mov

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203922830586217

### CODE REVIEW

**Comment by nonparibus** [comment]
@yusufkinatas A couple of observations:

1. we highlight the active column when you click in a cell, but not the active row. Could we do that, too?
1. can we have the 'New row' row flush with the row above it? It's a bit odd to me that there's a partial gap between.

**Comment by yusufkinatas** [comment]
@nonparibus Addressed all of the comments!

**Review by nonparibus — DISMISSED** [review]
Legal approval ✅

Code review pending

Functional review pending

**Comment by nathggns** [comment]
@nonparibus @CiaranMn what's the plan for this one?

**Comment by nonparibus** [comment]
@nathggns OOS for your updates. Some YX fixes required. We're going to revisit later in the week, or early the following.

File: blocks/table-v2/src/components/settings/settings.tsx
Comment (nonparibus) [inline blocks/table-v2/src/components/settings/settings.tsx:66]: We should expand the pop-up so this text doesn't bleed onto two lines.

File: blocks/table-v2/src/app.tsx
Comment (nonparibus) [inline blocks/table-v2/src/app.tsx:190]: Once a single checkbox is selected, we should then display checkboxes next to all other rows (in unchecked state, unless/until flipped).

We also want to allow for multiple checkboxes to then be selected, without having to shift+click or cmd+click (if not already supported).

File: blocks/table-v2/src/components/grid/grid.tsx
Comment (nonparibus) [inline blocks/table-v2/src/components/grid/grid.tsx:9]: Wasn't sure where to drop this comment... below a certain number of rows, I believe it is still possible to get left with a weird partial row between the final actual row of cells, and the "new row" row. We should make sure this is flush always.

Capture 2023-03-09 at 17 51 50@2x

File: blocks/table-v2/LICENSE-APACHE.md
Comment (nonparibus) [inline blocks/table-v2/LICENSE-APACHE.md:2]: As part of this PR, please remove the old `blocks/table` folder and replace it with this (renaming this from `table-v2` to just `table` in the process).

We don't want the old table code (targeting an unsupported version of the spec) hanging around any more.

File: blocks/table-v2/src/app.tsx
Comment (yusufkinatas) [reply blocks/table-v2/src/app.tsx]: Re: displaying all checkboxes

https://github.com/glideapps/glide-data-grid/commit/91ad40b8bc39a22d98994530530b037083509ab9 `glide-data-grid` will add a row-marker option called `checkbox-visible`  on their next patch (`5.2.2`). I suggest waiting until they release this (they updated their repo 9days ago, I guess they’ll release new patch soon)

File: blocks/table-v2/LICENSE-APACHE.md
Comment (yusufkinatas) [reply blocks/table-v2/LICENSE-APACHE.md]: https://github.com/hashintel/hash/pull/1946/commits/2048c0c2395c40a93e49db706cb444763e379794 done here

File: blocks/table-v2/src/components/grid/grid.tsx
Comment (yusufkinatas) [reply blocks/table-v2/src/components/grid/grid.tsx]: Addressed this issue here https://github.com/hashintel/hash/pull/1946/commits/13b8a7de657b465ab8f27366cc7887841936dfb6

File: blocks/table-v2/src/app.tsx
Comment (yusufkinatas) [reply blocks/table-v2/src/app.tsx]: Addressed the other issue here https://github.com/hashintel/hash/pull/1946/commits/13b8a7de657b465ab8f27366cc7887841936dfb6

File: blocks/table-v2/src/components/settings/settings.tsx
Comment (yusufkinatas) [reply blocks/table-v2/src/components/settings/settings.tsx]: https://github.com/hashintel/hash/pull/1946/commits/13b8a7de657b465ab8f27366cc7887841936dfb6 expanded it here

File: blocks/table/.gitignore
Comment (CiaranMn) [inline blocks/table/.gitignore:3]: these are covered by the root `.gitignore`, you can delete this

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:3]: suggest making this `0.0.1`

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:19]: Covered by root formatter
```suggestion
```

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:22]: These should be here and you should keep the `eslintrc.cjs` file

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:40]:
```suggestion
```

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:36]: These should remain

File: blocks/table/package.json
Comment (CiaranMn) [inline blocks/table/package.json:39]: This should remain

File: blocks/table/src/app.tsx
Comment (CiaranMn) [inline blocks/table/src/app.tsx:18]: when you restore linting these should be autosorted

File: blocks/table/src/app.tsx
Comment (CiaranMn) [inline blocks/table/src/app.tsx:114]: if we `expect-error` we will be warned if this line becomes unnecessary – also suggest mentioning what error we are expecting in case more appear when someone revisits this
```suggestion
          // @ts-expect-error -- type instantiation is deep and possibly infinite
```

File: blocks/table/src/app.tsx
Comment (CiaranMn) [inline blocks/table/src/app.tsx:129]: `updateEntity` is defined after it is used, I'd move this up (linting will probably flag this)

File: blocks/table/src/app.tsx
Comment (CiaranMn) [inline blocks/table/src/app.tsx:299]: what's this for? is it a `@todo`?
```suggestion
              // update column title here
```

**Review by CiaranMn — COMMENTED** [review]
Thanks Yusuf – tested this out and works well.

I've a few comments, mainly around making sure this has the same linting and commands as other blocks in the repo – once you reactive linting there might be a few other things to fix, hopefully mostly autofixable

File: blocks/table/package.json
Comment (yusufkinatas) [reply blocks/table/package.json:3]: Since previous version of the table block was `0.2.0`, could making it `0.2.1` makes more sense?

Comment (CiaranMn) [reply blocks/table/package.json:3]: Eh, I lean towards no, because this is really just a new block that happens to be taking over the name of the old one. It has no relation to it other than the name. I don't feel strongly about it though – maybe @nonparibus has a preference

Comment (yusufkinatas) [reply blocks/table/package.json:3]: I thought there might be a technical reason to always increase the version while using the same `name` for the package (everything is the same, directory, package name etc.)

Comment (CiaranMn) [reply blocks/table/package.json:3]: If anyone was using / relying on the old one I think there would be, but they aren't at the moment. I think restarting the version gives a more accurate description of the development history of this specific block – i.e. it's just started

Comment (nonparibus) [reply blocks/table/package.json:3]: I would personally increment the version number still. I think it is extremely confusing not to do so. A whole rewrite might be a v1 -> v2 jump in ordinary product-land, but with this being pre-v1 I would just go 0.2 -> 0.3

I absolutely don't think the version number should _decrease_ while using the same reserved block name. Recipe for confusion.

Comment (nonparibus) [reply blocks/table/package.json:3]: Please make this `0.3`

**Comment by yusufkinatas** [comment]
@CiaranMn Addressed all of the comments and fixed some eslint errors here https://github.com/hashintel/hash/pull/1946/commits/4668257c96693f4e8c19bed6db273ecac15bc3f4

**Comment by yusufkinatas** [comment]
I guess next step before merging this is, if it gets approved we should create the block entity type at BP production env, then merge this?

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**
